lib/fluent/plugin/out_datadog.rb

```diff
@@ -131,40 +108,38 @@ def write(chunk)
         messages.push "#{api_key} " + Yajl.dump(record) + "\n"
       else
         next unless record.has_key? "message"
-        messages.push "#{api_key} " + record["message"].rstrip() + "\n"
+        messages.push "#{api_key} " + record["message"].rstrip + "\n"
       end
     end
     send_to_datadog(messages)
 
   end
 
-  def send_to_datadog(data)
+  def send_to_datadog(events)
     @my_mutex.synchronize do
-      retries = 0
-      begin
-        log.trace "Send nb_event=#{data.size} events to Datadog"
-
-        # Check the connectivity and write messages
-        log.info "New attempt to Datadog attempt=#{retries}" if retries > 0
-
-        data.each do |event|
-          log.trace "Datadog plugin: about to send event=#{event}"
-          client.write(event)
-        end
-
-        # Handle some failures
-      rescue Errno::EHOSTUNREACH, Errno::ECONNREFUSED, Errno::ETIMEDOUT, Errno::EPIPE => e
-
-        if retries < @max_retries || max_retries == -1
-          @_socket = nil
-          a_couple_of_seconds = retries ** 2
-          a_couple_of_seconds = 30 unless a_couple_of_seconds < 30
-          retries += 1
-          log.warn "Could not push logs to Datadog, attempt=#{retries} max_attempts=#{max_retries} wait=#{a_couple_of_seconds}s error=#{e.message}"
-          sleep a_couple_of_seconds
-          retry
+      log.trace "Sending nb_event=#{events.size} events to Datadog"
+
+      events.each do |event|
+        log.trace "Datadog plugin: about to send event=#{event}"
+        retries = 0
+        begin
+          log.info "New attempt to Datadog attempt=#{retries}" if retries > 0
+          @client ||= new_client
+          @client.write(event)
+        rescue => e
+          if retries < @max_retries || max_retries == -1
+            # Restart a new connection
+            @client.close rescue nil
+            @client = nil
+            a_couple_of_seconds = retries ** 2
+            a_couple_of_seconds = 30 unless a_couple_of_seconds < 30
+            retries += 1
+            log.warn "Could not push event to Datadog, attempt=#{retries} max_attempts=#{max_retries} wait=#{a_couple_of_seconds}s error=#{e}"
```

Comment:
should we log only if the first retry is not successful? To avoid logging when the connection has been closed server side and we just need to open a new one?

Reply:
I am not sure we should stop logging this error

Reply:
Given that just about the only time I see this message is on connections that have been closed server-side, I would love to skip logging the first one, and only see it if there's a sustained connection problem. Also maybe we should close the socket regardless of whether we'll retry or not? Something like this, maybe?

```ruby
rescue => e
  @client.close rescue nil
  @client = nil
  retries += 1

  retry if retries == 1 # immediately retry, in case it's just a server-side close

  if retries < @max_retries || @max_retries == -1
    a_couple_of_seconds = Integer.min(retries ** 2,  30)
    log.warn "Could not push event to Datadog, attempt=#{retries} max_attempts=#{max_retries} wait=#{a_couple_of_seconds}s error=#{e}"
    sleep a_couple_of_seconds
    retry
  end
  raise ConnectionFailure, "Could not push event to Datadog after #{retries} retries, #{e}"
end

Reply:
Maybe a compromise would be to log the first connection failure at `trace`?